### PR TITLE
Ctrl-C management in here_doc plus print echo $?

### PR DESCRIPTION
**Pull request by dbejar-s**

### CODE REVIEW

File: src/general_utils.c [hunk 1]
```diff
@@ -94,5 +94,8 @@ t_macro	*init_macro(char **envp, char **argv)
 	macro->pipe_fd[0] = -1;
 	macro->pipe_fd[1] = -1;
 	macro->read_end = -1;
+	macro->exit_code = 0;
```
Comment (pmarkaide) [inline src/general_utils.c:97]: Not needed to set to zero if we calloc, supposatelly

File: src/main.c [hunk 2]
```diff
@@ -20,7 +20,18 @@ static char	*read_line(t_macro *macro)
 	char	*path;
 
 	path = create_path(macro);
+	if (macro->here_doc_flag == 1)
+	{
+		signal(SIGINT, sigint_handler_after_here_doc);
+		macro->here_doc_flag = 0;
+	}
 	line = readline(path);
+	signal(SIGINT, sigint_handler_in_parent);
```
Comment (pmarkaide) [inline src/main.c:29]: Nice

File: src/signals.c [hunk 3]
```diff
@@ -39,3 +40,21 @@ void	sigquit_handler_in_child(int sig)
 		ft_printf("Quit\n");
 	}
 }
+
+void	sigint_handler_after_here_doc(int sig)
```
Comment (pmarkaide) [inline src/signals.c:44]: Isn't this just the same as siging_handler_in_parent?

File: src/parsing_utils.c [hunk 4]
```diff
@@ -48,7 +51,21 @@ static int	read_here_doc(t_token *token, t_macro *macro)
 	del = clean_quotes(token->value);
 	while (1)
 	{
+		signal(SIGINT, sigint_handler_here_doc);
 		line = readline("> ");
+		if (g_exit == SIGINT)
+		{
+			open("/dev/tty", O_RDONLY);
```
Comment (pmarkaide) [inline src/parsing_utils.c:58]: can you explain what /dev/tty does?

File: src/signals.c [hunk 5]
```diff
@@ -39,3 +40,21 @@ void	sigquit_handler_in_child(int sig)
...
+		rl_redisplay();
+		g_exit = sig;
+	}
+}
+
+void	sigint_handler_here_doc(int sig)
+{
+	if (sig == SIGINT)
+	g_exit = SIGINT;
+	close(STDIN_FILENO);
```
Comment (pmarkaide) [inline src/signals.c:59]: This is clever, I couldn't stop the deadline sending EOF or anything

**Review by pmarkaide — COMMENTED** [review]
I've tested and it works. Is clear and I like the implementation design. We need to add a newline also in the here_doc_handler to make it as bash, but apart from that seem to work. Let's discuss details in the meeting and we merge

File: src/signals.c (hunk 5 above)
Comment (dbejar-s) [reply src/signals.c:59]: I tried so many different ways to make readline stops... and I knew it needs EOF and the only way I figured out to "fake" a EOF is by closing the STDIN. Problem is that you need to open the STDIN as soon as you're done with the exit procedure caused by SIGINT, hence the open("/dev/tty", O_RDONLY)... which is basically a way to undo the close(STDIN_FILENO). Thing is that opening /dev/tty ignores any other fd redirection that is in place... but this only happens with here_doc, so it should be ok

File: src/signals.c (hunk 3 above)
Comment (dbejar-s) [reply src/signals.c:44]: Difference is that I don't print a new line... I just copied my bash behavior. Let's check it out.

File: src/signals.c (hunk 5 above)
Comment (pmarkaide) [reply src/signals.c:59]: I tried using `ioctl(STDIN_FILENO, TIOCSTI, "\004");` that sends EOF to the terminal, but it did not close deadline. I will ask in school

**Pull request merged**